env/env_test.cc

```diff
@@ -2644,6 +2645,24 @@ TEST_F(EnvTest, GenerateRawUniqueIdTrackRandomDeviceOnly) {
   t.Run();
 }
 
+TEST_F(EnvTest, FailureToCreateLockFile) {
+  auto env = Env::Default();
+  auto fs = env->GetFileSystem();
+  std::string dir = test::PerThreadDBPath(env, "lockdir");
+  std::string file = dir + "/lockfile";
+
+  ASSERT_OK(DestroyDir(env, dir));
+
+  // Make sure that we can acquire a file lock after the first attempt fails
+  FileLock* lock = nullptr;
+  ASSERT_NOK(fs->LockFile(file, IOOptions(), &lock, /*dbg*/ nullptr));
+  ASSERT_FALSE(lock);
+
+  ASSERT_OK(fs->CreateDir(dir, IOOptions(), /*dbg*/ nullptr));
+  ASSERT_OK(fs->LockFile(file, IOOptions(), &lock, /*dbg*/ nullptr));
+  delete lock;
```

Comment:
Shouldn't this call UnlockFile to remove the file from the lock list rather than deleting the lock?

Also, should this DestroyDir at the end to clean up after itself?

Reply:
Tests can't rely on a pristine initial environment because prior tests might have failed and cascading failures are confusing. It's simpler to just clean up at the beginning.

Reply:
Do you clean out your dryer lint before and after each cycle? ;)

Reply:
And why is FileLock a defined class in the public API if I'm not supposed to call delete on it? I saw it had a virtual destructor and inferred what to do with it.

Reply:
> And why is FileLock a defined class in the public API if I'm not supposed to call delete on it?

For custom FileSystem implementation

Reply:
FileLock should be a struct not a class.  Not something you can create or destroy on your own.  Water under the bridge...

Reply:
> Tests can't rely on a pristine initial environment because prior tests might have failed and cascading failures are confusing. It's simpler to just clean up at the beginning.

I agree that tests cannot rely on a pristine environment but it would be nice if we did not leave a bunch of turds around.  And since the next time through this could generate a different directory, I can see many dingleberries piling up...

Reply:
I thought the framework deleted these directories on exit but I guess that's only dbname_. I'll fix.